File: advanced-programming-methods/toy-interpreter-java/src/repository/MemRepository.java
package repository;

import model.ProcTable;
import model.ProgramState;
import model.statement.Statement;

import java.io.*;
import java.util.ArrayList;

/**
 * The type Mem repository.
 */
public class MemRepository implements IRepository
{
    private ArrayList<ProgramState> programs            = new ArrayList<ProgramState>();
    private int                     currentProgramIndex = 0;
    private ProgramState currentProgram;
    private ProcTable procTable = new ProcTable();

    // Automatic program ID set.
    private int availableID = 0;

    private static final String filename_serialized = "program.toy.ser";
    private static final String filename            = "program.toy";

    /**
     * Instantiates a new Mem repository.
     *
     * @param firstProgram the program state
     */
    public MemRepository(ProgramState firstProgram)
    {
        currentProgram = firstProgram;
        setProgramID(currentProgram);
    }

    /**
     * Getter for property 'procTable'.
     *
     * @return Value for property 'procTable'.
     */
    public ProcTable getProcTable()
    {
        return procTable;
    }

    private void setProgramID(ProgramState program)
    {
        if (program.getID() == -1)
        {
            program.setID(availableID);
            availableID++; // next available one
        }
    }

    @Override
    public void addProgram(ProgramState program)
    {
        setProgramID(program);
        programs.add(program);
    }

    @Override
    public void removeProgram(ProgramState program)
    {
        programs.remove(program);
    }

    @Override
    public boolean nextCurrentProgram()
    {
        if (!hasPrograms())
        {
            return false;
        }

        // can execute a next program
        if (currentProgramIndex + 1 < programs.size())
        {
            // move forward
            currentProgramIndex++;
            currentProgram = programs.get(currentProgramIndex);
            return true;
        }

        // reset
        currentProgramIndex = 0;
        currentProgram = programs.get(currentProgramIndex);
        return false;
    }

    /** {@inheritDoc} */
    public void setMain(Statement statement)
    {
        currentProgram.setProgram(statement);
    }

    /** {@inheritDoc} */
    public boolean startMain() // it is called only once
    {
        assert programs.size() == 0;
        assert currentProgramIndex == 0;
        addProgram(currentProgram);

        return currentProgram.start();
    }

    public void stopMain()
    {
        programs.clear();
        currentProgramIndex = 0;

        currentProgram.stop(true);
    }

    /**
     * {@inheritDoc}
     */
    public ProgramState getCP()
    {
        return currentProgram;
    }

    /** {@inheritDoc} */
    public void stopCP() // step every program
    {
        removeProgram(currentProgram);

        currentProgramIndex--; // reduce index, we remove a program
        if (currentProgramIndex < 0)
        {
            currentProgramIndex = 0;
        }

        currentProgram.stop(false);
    }

    public boolean hasPrograms()
    {
        return programs.size() != 0;
    }

    @Override
    public boolean serializeToFile()
    {
        if (currentProgram == null)
        {
            return false;
        }

        try
        {
            FileOutputStream fileOut = new FileOutputStream(filename_serialized);
            ObjectOutputStream out = new ObjectOutputStream(fileOut);

            out.writeObject(currentProgram);

            out.close();
            fileOut.close();
            return true;
        }
        catch (IOException e)
        {
            System.out.println("ERROR: could not write: " + e.getMessage());
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public boolean serializeFromFile()
    {
        try
        {
            FileInputStream fileIn = new FileInputStream(filename_serialized);
            ObjectInputStream in = new ObjectInputStream(fileIn);

            currentProgram = (ProgramState) in.readObject();

            in.close();
            fileIn.close();
            return true;
        }
        catch (IOException e)
        {
            System.out.println("ERROR: could not read: " + e.getMessage());
            e.printStackTrace();
        }
        catch (ClassNotFoundException e)
        {
            System.out.println("ERROR: could not find class: " + e.getMessage());
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public boolean saveToFile()
    {
        if (currentProgram == null)
        {
            return false;
        }

        try
        {
            PrintWriter writer = new PrintWriter(filename);

            writer.println("// ExeStack content:");
            writer.println(currentProgram.getStack().getAllSave());

            writer.println("// SymTable content:");
            writer.println(currentProgram.getTable().peek().getAllSave());

            writer.println("// Out content:");
            writer.print(currentProgram.getOut().getAll());

            writer.close();
            return true;
        }
        catch (IOException e)
        {
            System.out.println("ERROR: could not write: " + e.getMessage());
            e.printStackTrace();
        }

        return false;
    }

    @Override
    public String toString()
    {
        String out = "";

        if (programs.size() > 1)
        {
            for (int i = 1; i < programs.size(); i++)
            {
                out += String.format("%s\n", programs.get(i));
            }
        }
        out += String.format("currentProgram = %s\n", currentProgram);
        out += String.format("procTable = %s", procTable.toString());

        return out;
    }
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/repository/MemRepository.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Runtime.Serialization;
using System.Collections.Generic;
using System.Runtime.Serialization.Formatters.Binary;
using Exceptions;
using Model;

namespace Repository
{
    /// <summary>
    /// The type Mem repository.
    /// </summary>
    public class MemRepository : IRepository
    {
        private ProgramState currentProgram;
        private List<ProgramState> programs = new List<ProgramState>();
        private int currentProgramIndex = 0;
        private int availableID = 0;

        private const string filename_serialized = "program.toy.ser";
        private const string filename = "program.toy";

        private void SetProgramID(ProgramState program)
        {
            if (program.GetID() == -1)
            {
                program.SetID(availableID);
                availableID++; // next available
            }
        }

        /// <summary>
        /// Instantiates a new Mem repository.
        /// </summary>
        /// <param name="programState"> the program state </param>
        public MemRepository(ProgramState mainProgram)
        {
            this.currentProgram = mainProgram;
            SetProgramID(currentProgram);
        }

        public void AddProgram(ProgramState program)
        {
            programs.Add(program);
            SetProgramID(program);
        }

        public void RemoveProgram(ProgramState program)
        {
            programs.Remove(program);
        }

        public bool NextCurrentProgram()
        {
            if (!HasPrograms())
            {
                return false;
            }

            // can execute a next program
            if (currentProgramIndex + 1 < programs.Count)
            {
                // move forward
                currentProgramIndex++;
                currentProgram = programs[currentProgramIndex];
                return true;
            }

            // reset
            currentProgramIndex = 0;
            currentProgram = programs[currentProgramIndex];
            return false;
        }

        /// <inheritdoc />
        public void SetMain(Statement statement)
        {
            currentProgram.SetProgram(statement);
        }

        /// <inheritdoc />
        public bool StartMain()
        {
            Debug.Assert(programs.Count == 0);
            Debug.Assert(currentProgramIndex == 0);
            AddProgram(currentProgram);
            currentProgramIndex = 0;

            return currentProgram.Start();
        }

        public void StopMain()
        {
            programs.Clear();
            currentProgramIndex = 0;

            currentProgram.Stop(true);
        }

        public ProgramState GetCP()
        {
            return currentProgram;
        }

        /// <inheritdoc />
        public void StopCP()
        {
            RemoveProgram(currentProgram);

            currentProgramIndex--; // reduce index, we remove a program
            if (currentProgramIndex < 0)
            {
                currentProgramIndex = 0;
            }

            currentProgram.Stop(false);
        }

        /// <inheritdoc />
        public bool HasPrograms()
        {
            return programs.Count != 0;
        }

        public bool SerializeToFile()
        {
            if (currentProgram == null)
            {
                return false;
            }

            try
            {
                FileStream fs = File.Create(filename_serialized);

                BinaryFormatter bformatter = new BinaryFormatter();

                bformatter.Serialize(fs, currentProgram);

                fs.Close();
                return true;
            }
            catch (SerializationException e)
            {
                Console.WriteLine("ERROR: could not write: " + e.Message);
            }
            catch (IOException e)
            {
                Console.WriteLine("ERROR: could not write: " + e.Message);
            }
          
            return false;
        }

        public bool SerializeFromFile()
        {
            try
            {
                FileStream fs = File.OpenRead(filename_serialized);
                BinaryFormatter bformatter = new BinaryFormatter();

                currentProgram = (ProgramState) bformatter.Deserialize(fs);

                fs.Close();
                return true;
            }
            catch (SerializationException e)
            {
                Console.WriteLine("ERROR: could not read: " + e.Message);
            }
            catch (IOException e)
            {
                Console.WriteLine("ERROR: could not read: " + e.Message);
            }

            return false;
        }

        public bool SaveToFile()
        {
            if (currentProgram == null)
            {
                return false;
            }

            try
            {
                StreamWriter writer = new StreamWriter(filename);

                //writer.WriteLine(this.ToString());

                writer.WriteLine("// ExeStack content:");
                writer.WriteLine(currentProgram.Stack.AllSave);

                writer.WriteLine("// SymTable content:");
                writer.WriteLine(currentProgram.Table.AllSave);

                writer.WriteLine("// Out content:");
                writer.Write(currentProgram.Out.All);

                writer.Close();
                return true;
            }
            catch (IOException e)
            {
                Console.WriteLine("ERROR: could not save: " + e.Message);
            }

            return false;
        }

        /// <inheritdoc />
        public override string ToString()
        {
            string ret = "";

            if (programs.Count > 1)
            {
                for (int i = 1; i < programs.Count; i++)
                {
                    ret += string.Format("{0}\n", programs[i].ToString());
                }
            }
            ret += string.Format("currentProgram = {0}", currentProgram.ToString());

            return ret;
        }
    }

}

